test/micro/org/openjdk/bench/vm/compiler/VectorAlgorithmsImpl.java

```diff
@@ -0,0 +1,418 @@
+        int index = idxs.blend(a.length, not_min_mask).reduceLanes(VectorOperators.MIN);
+        // Tail loop
+        for (; i < a.length; i++) {
+            int ai = a[i];
+            if (ai < min) {
+                min = ai;
+                index = i;
+            }
+        }
+        return index;
+    }
+
+    public static int findI_loop(int[] a, int e) {
+        for (int i = 0; i < a.length; i++) {
+            int ai = a[i];
+            if (ai == e) {
+                return i;
+            }
+        }
+        return -1;
+    }
+
+    public static int findI_VectorAPI(int[] a, int e) {
+        var es = IntVector.broadcast(SPECIES_I, e);
+        int i = 0;
+        for (; i < SPECIES_I.loopBound(a.length); i += SPECIES_I.length()) {
+            IntVector v = IntVector.fromArray(SPECIES_I, a, i);
+            var mask = v.compare(VectorOperators.EQ, es);
+            if (mask.anyTrue()) {
+                var ml = mask.toLong();
+                return i + Long.numberOfTrailingZeros(ml);
+            }
+        }
+        for (; i < a.length; i++) {
+            int ai = a[i];
+            if (ai == e) {
+                return i;
+            }
+        }
+        return -1;
+    }
+
+    public static Object reverseI_loop(int[] a, int[] r) {
+        for (int i = 0; i < a.length; i++) {
+            r[a.length - i - 1] = a[i];
+        }
+        return r;
+    }
+
+    private static final VectorShuffle<Integer> REVERSE_SHUFFLE_I = SPECIES_I.iotaShuffle(SPECIES_I.length()-1, -1, true);
+
+    public static Object reverseI_VectorAPI(int[] a, int[] r) {
+        int i = 0;
+        for (; i < SPECIES_I.loopBound(a.length); i += SPECIES_I.length()) {
+            IntVector v = IntVector.fromArray(SPECIES_I, a, i);
+            v = v.rearrange(REVERSE_SHUFFLE_I);
+            v.intoArray(r, r.length - SPECIES_I.length() - i);
+        }
+        for (; i < a.length; i++) {
+            r[a.length - i - 1] = a[i];
+        }
+        return r;
+    }
+
+    public static Object filterI_loop(int[] a, int[] r, int threshold) {
+        int j = 0;
+        for (int i = 0; i < a.length; i++) {
+            int ai = a[i];
+            if (ai >= threshold) {
+                r[j++] = ai;
+            }
+        }
+        // Just force the resulting length onto the same array.
+        r[r.length - 1] = j;
+        return r;
+    }
+
+    public static Object filterI_VectorAPI(int[] a, int[] r, int threshold) {
+        var thresholds = IntVector.broadcast(SPECIES_I, threshold);
+        int j = 0;
+        int i = 0;
+        for (; i < SPECIES_I.loopBound(a.length); i += SPECIES_I.length()) {
+            IntVector v = IntVector.fromArray(SPECIES_I, a, i);
+            var mask = v.compare(VectorOperators.GE, thresholds);
+            v = v.compress(mask);
+            int trueCount = mask.trueCount();
+            var prefixMask = VectorMask.fromLong(SPECIES_I, (1L << trueCount) - 1);
```

Comment:
In case you did not already try, you can compress the mask to produce the mask prefix, as in `v.intoArray(r, j, mask.compress())`. Unsure if that will make a difference in the codegen though.

Reply:
Ah, nice idea!